acra/configuring-maintaining/optimisations/acrastructs_vs_acrablocks.md

```diff
@@ -0,0 +1,44 @@
+---
+title: AcraStructs vs AcraBlocks
+weight: 3
+---
+
+# AcraStructs vs AcraBlocks
+
+There are two different "crypto containers" AcraServer can use to store encrypted value in.
+Then, these containers are stored in database and decrypted back by AcraServer when needed.
+
+`AcraStruct` is the main crypto container for the client-side encryption.
+Applications can encrypt with ephemeral asymmetric keys on their side with `AcraStruct`s,
+but cannot do this with `AcraBlock`s because of symmetric key encryption.
+You can also configure AcraServer to re-encrypt `AcraStruct`s (that were generated on app side)
+on AcraServer side, this way data will definitely be encrypted while transmitted between
+some client app and AcraServer.
```

Comment:
> this way data will definitely be encrypted while transmitted between some client app and AcraServer.

It's it already encrypted by the transport layer in flight in case of transparent encryption?

I think the point of reencrypting AcraStructs is to make sure that you can't match anything known to the application to the data stored in the database. So an attacker can't submit some AcraStruct, expect it to be stored verbatim, then peek into the database for what row it has been stored at, and then learn something about the system from that.

Another point is preventing usage of compromised keys. AcraServer might accept AcraStructs encrypted with old keys (not necessarily compromised but rotated out), and then reencrypt the data with the current ones.

@Lagovas, @storojs72, what to you think about this part of the paragraph?

Reply:
> You can also configure AcraServer to re-encrypt AcraStructs (that were generated on app side) on AcraServer side, this way data will definitely be encrypted while transmitted between some client app and AcraServer

I agree, it sounds a bit odd. If AcraStruct came to AcraServer from client app, then plaintext from app is definitely encrypted - so there is no need to additionally ensure that data is encrypted. @ilammy's explanation sounds more relevant - with mentioned re-encryption attacker is not able to push specifically crafted AcraStructs to AcraServer and then observe them in the database, since after re-encryption AcraStruct will have different value.

Reply:
+1, you are right